src/motion_pvt_sequence_generation/pvt.py

```diff
@@ -0,0 +1,957 @@
+
+    See https://en.wikipedia.org/wiki/Cubic_Hermite_spline#Finite_difference.
+
+    :param position_sequence: An array containing the position before, at, and after, the
+        generation point.
+    :param time_sequence: An array containing the time before, at, and after, the generation
+        point.
+    :return: The interpolated velocity.
+    """
+    # Setup
+    sum_differences = 0.0
+    num_differences = 0.0
+
+    # Calculate the finite differences for the two segments involving the
+    # point at which to generate a velocity.
+    for index in range(2):
+        delta_time = time_sequence[index + 1] - time_sequence[index]
+        if delta_time > 0:
+            num_differences += 1
+            sum_differences += (
+                position_sequence[index + 1] - position_sequence[index]
+            ) / delta_time
+    assert num_differences > 0, "All three points must not have the same time"
+
+    # Return the average difference
+    return sum_differences / num_differences
+
+
+class CSVData:
+    """A helper class to read sequences from CSV files."""
+
+    _time_index: int | None
+    """The index of the time column."""
+    _position_indices: list[int]
+    """The indices of the position columns."""
+    _velocity_indices: list[int]
+    """The indices of the position columns."""
+    _time_sequence: list[float]
+    """The sequence of time values read from the file."""
+    _position_sequences: list[list[float]]
+    """An array of position sequences, one for each dimension"""
+    _velocity_sequences: list[list[float | None]]
+    """An array of velocity sequences, one for each dimension."""
+
+    @property
+    def contains_time_data(self) -> bool:
+        """Return whether the data contains time values."""
+        return self._time_index is not None and any(t is not None for t in self._time_sequence)
+
+    @property
+    def contains_position_data(self) -> bool:
+        """Return whether the data contains position values."""
+        if len(self._position_indices) > 0:
+            for sequence in self._position_sequences:
+                if any(p is not None for p in sequence):
+                    return True
+        return False
+
+    @property
+    def contains_velocity_data(self) -> bool:
+        """Return whether the data contains velocity values."""
+        if len(self._velocity_indices) > 0:
+            for sequence in self._velocity_sequences:
+                if any(v is not None for v in sequence):
+                    return True
+        return False
+
+    @property
+    def contains_complete_velocity_data(self) -> bool:
+        """Return whether or not all velocity values are specified."""
+        if len(self._velocity_indices) == 0:
+            return False
+        for sequence in self._velocity_sequences:
+            if any(v is None for v in sequence):
+                return False
+        return True
+
+    @property
+    def time_sequence(self) -> list[float] | None:
```

Comment:
You're right, this is a better way of doing it. I originally added the `None` return path to make it explicit that no data of the given type was read from the file, but this can be better achieved with an assert. See 6c4f3a4c50ec3543b53d99d792fc1a831bd8b120. There are still two `type: ignore`s that I can't get rid of due to the fact that velocity data can be incomplete.